articles/using_synapse_as_an_oauth_server.md

```diff
@@ -0,0 +1,129 @@
+---
+title: "Using Synapse as an OAuth 2.0 Server"
+layout: article
+excerpt: Follow these steps to register an OAuth Client and authorize access to Synapse.
+category: howto
+---
+
+<style>
+#image {
+    width: 100%;
+}
+#imageSmall {
+    width: 40%;
+}
+</style>
+
+
+# Registering and linking an OAuth 2.0 Client
+Synapse implements the [OpenID Connect](https://openid.net/specs/openid-connect-core-1_0.html) protocol, which is an extension of OAuth 2.0.  The Open ID Connect endpoint is `https://repo-prod.prod.sagebase.org/auth/v1`, and the Open ID Configuration (aka the "discovery document") is published in the standard location: [https://repo-prod.prod.sagebase.org/auth/v1/.well-known/openid-configuration](https://repo-prod.prod.sagebase.org/auth/v1/.well-known/openid-configuration).
+
+## Create an OAuth 2.0 Client
+You can create a client programmatically as follows:
+
+
+##### Python
+
+```python
+import synapseclient
+import json
+syn = synapseclient.login()
+
+clientMetaData = {
+  'client_name': 'Your client name',
+  'redirect_uris': [
+    'https://yourhost.com/user/login'
+  ],
+  'client_uri': 'https://yourhost.com/index.html',
+  'policy_uri': 'https://yourhost.com/policy',
+  'tos_uri': 'https://yourhost.com/terms_of_service',
+  'userinfo_signed_response_alg': 'RS256'
+}
+
+clientMetaData = syn.restPOST(uri='/oauth2/client', endpoint=syn.authEndpoint, body=json.dumps(clientMetaData))
+
+client_id = clientMetaData['client_id']
+
+clientIdAndSecret = syn.restPOST(uri='/oauth2/client/secret/'+client_id, endpoint=syn.authEndpoint, body='')
+
+print(clientIdAndSecret)
+
+```
+
+The client URI, policy URI, and terms of service URI are optional, as is the `userinfo_signed_response_alg` parameter.  Further, you can optionally include a `sector_identifier_uri` parameter.  This is an advanced feature described [here](https://openid.net/specs/openid-connect-registration-1_0.html#SectorIdentifierValidation) and is relevant if the client uses multiple hosts because Synapse only returns `pairwise` subject values to its OAuth clients.
+
+
+The returned `client_id` and `client_secret` will be needed later when getting an access token.  The secret is only returned once.  (Synapse does not keep it.)  If lost, you can generate a new secret but the previous one will be invalidated.
+
+You can retrieve, update, and delete your client programmatically as well:
+
+```python
+clientMetaData = syn.restGET(uri='/oauth2/client/'+client_id, endpoint=syn.authEndpoint)
+
+clientMetaData['policy_uri'] = 'https://yourhost.com/updated_policy'
+
+clientMetaData = syn.restPUT(uri='/oauth2/client/'+client_id, endpoint=syn.authEndpoint, clientMetaData)
+
+syn.restDELETE(uri='/oauth2/client/'+client_id, endpoint=syn.authEndpoint)
```

Comment:
If this returns a value it should be captured for later inspection.

Reply:
it doesn't